src/cryptography/hazmat/backends/openssl/dh.py

```diff
@@ -261,6 +261,20 @@ def __init__(self, backend: Backend, dh_cdata, evp_pkey):
     def key_size(self) -> int:
         return self._key_size_bits
 
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, dh.DHPublicKey):
+            return NotImplemented
+
+        res = self._backend._lib.EVP_PKEY_cmp(
+            self._evp_pkey, other._evp_pkey  # type: ignore[attr-defined]
+        )
```

Comment:
We should either type-check against the concrete class, or we should make the implementation use exclusively the public API.